Source/Core/ProgramElement.cs

```diff
@@ -0,0 +1,103 @@
+using System;
+
+namespace Microsoft.Boogie
+{
+  public abstract record ProgramElement()
```

Comment:
Subjective: my preference for the name would be `TrackedNodeComponent` and for the subclasses to be `TrackedCallRequires`, `TrackedInvariantMaintained` etc. The `Node` part of the name than overlaps with `Call` and `Invariant` while `Component` overlaps with `RequiresGoal`, `RequiresAssumed` and `Ensures` for `Call`, and `Assumed`, `Established`, `Maintained` for `Invariant`.